README.md
# MSFS Landing Inspector
MSFS Landing Inspector is a tool for analyzing landings in Microsoft Flight Simulator 2020. It reads the airplane’s telemetry data via SimConnect and displays relevant information about your landing in a web browser. The MSFS Landing Inspector is free to use.

## October 7, 2020: Update Version 1.2:
- **Improved G force values at touchdown.**
- **Added new data readings and graphs for airspeed, pitch, bank, heading, wind speed/direction.**
- **Added custom icons for HTML site.**

### MSFS Landing Inspector displays the following data:
- Current vertical G force
- Current vertical speed
- Touchdown G force
- Touchdown vertical speed
- Touchdown airspeed
- Touchdown pitch, bank and heading
- Touchdown wind speed and direction
- Graph showing G forces during landing
- Graph showing vertical speed during landing
- Graph showing altitude above ground during landing
- Graph showing airspeed during landing
- Graph showing pitch, bank and heading during landing
- Graph showing wind speed and direction during landing

Screenshot of MSFS Landing Inspector in action:
![](images/MSFS_Landing_Inspector_Screenshot.png)

## Requirements
-	Python 3.7 64-bit or later
-	Flask library

## Installation
1. Install the latest Python 3 64-bit Release for Windows. Download the installer from the official [Python website](https://www.python.org/downloads/windows/). Here is a direct link to the [Python 3.8.6 64-bit installer](https://www.python.org/ftp/python/3.8.6/python-3.8.6-amd64.exe). Tick the *Add Python 3.8 to PATH* checkbox when installing Python.
2. Run Command Prompt. Do this by pressing <kbd>Win</kbd> + <kbd>R</kbd>. Type *cmd.exe* and click OK.
3. In the Command Prompt window type *pip install flask* and press <kbd>Enter</kbd>. When asked to download any other Python dependencies, type *y* and press <kbd>Enter</kbd>. This will install the Flask Python library.

You can also download a standalone compiled executable version of MSFS Landing Inspector [here](https://github.com/mracko/MSFS-Landing-Inspector/releases/tag/1.2). The compiled executable doesn't require Python or any additional libraries.

## Running MSFS Landing Inspector
1. Start a flight in Microsoft Flight Simulator
2. Download and unzip the MSFS Landing Inspector repository and run *msfs_landing_inspector.py* by double-clicking on it. This should launch a Command Prompt window. Don't close this window.
3. Open your browser and load the site *localhost:5000*. This should load up the MSFS Landing Inspector in your browser.

## Running MSFS Landing Inspector on your mobile device
1. Make sure your PC and your mobile device are connected to the same local network and that your home network is set to *Private* in your Network Profile settings. 
2. Run Command Prompt. Do this by pressing <kbd>Win</kbd> + <kbd>R</kbd>.
3. In the Command Prompt window type *ipconfig* and press <kbd>Enter</kbd>. Look for the line *IPv4 Address* and write down the IP address. In my case, it's *192.168.0.120*. You will most likely have a different IP address. Close the Command Prompt window.
4. Start a flight in Microsoft Flight Simulator
5. Download and unzip the MSFS Landing Inspector repository and run *msfs_landing_inspector.py* by double-clicking on it. This should launch a Command Prompt window. Don't close this window.
6. On your mobile device, load the following site in the browser: *<IP address you've written down>:5000*. In my case, this would be *192.168.0.120:5000*. MSFS Landing Inspector should now load on your mobile device.

## Credits
MSFS Landing Inspector uses the Python [SimConnect](https://pypi.org/project/SimConnect/) library and the CSS theme [Black Dashboard](https://www.creative-tim.com/product/black-dashboard) by [Creative Tim](https://www.creative-tim.com/) for the web front-end.

## Donation
If you like this tool and would like to support the development, please consider donating by clicking on the link below.

[![paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=CXDDYFUSWA2Z4&source=url)


Happy landing!


CHANGELOG.md
# MSFS Landing Inspector Changelog

## Version 1.2 (October 7, 2020)
- Improved G force values at touchdown.
- Added new data readings (airspeed, pitch, bank, heading, wind speed/direction).
- Added custom icons for HTML site.

## Version 1.1 (October 1, 2020):
- Fixed -999999 values caused by low frame rates.
- Fixed vertical speed to represent true values instead of indicated.
- Landing rating is now based on the vertical speed at touchdown.
- Improved visualization of charts. Different colors are now shown for data when airborne and when on the ground.

## Version 1.0 (September 28, 2020)
- Initial release.
- Following data is displayed: current vertical G force, current vertical speed, touchdown G force, touchdown vertical speed, graph showing G forces during landing, graph showing vertical speed during landing, graph showing altitude above ground during landing


static/js/custom/getSimData.js
let altitude;
let fuel_percentage;
let vertical_speed;
let compass;
let airspeed;
let latitude;
let longitude;

let autopilot_master;
let autopilot_nav_selected;
let autopilot_wing_leveler;
let autopilot_heading_lock;
let autopilot_heading_lock_dir;
let autopilot_altitude_lock;
let autopilot_altitude_lock_var;
let autopilot_attitude_hold;
let autopilot_glidescope_hold;
let autopilot_approach_hold;
let autopilot_backcourse_hold;
let autopilot_vertical_hold;
let autopilot_vertical_hold_var;
let autopilot_pitch_hold;
let autopilot_pitch_hold_ref;
let autopilot_flight_director_active;
let autopilot_airspeed_hold;
let autopilot_airspeed_hold_var;

let gear_handle_position;
let elevator_trim_pct;
let elevator_trim_pct_reversed;
let rudder_trim_pct;
let flaps_handle_pct;
let flaps_handle_pct_reversed;

let cabin_seatbelts_alert_switch;
let cabin_no_smoking_alert_switch;

window.setInterval(function(){
    getSimulatorData();
    displayData()
    updateMap()
}, 100);


function getSimulatorData() {
    $.getJSON($SCRIPT_ROOT + '/ui', {}, function(data) {

        //Navigation
        altitude = data.ALTITUDE;
        vertical_speed = data.VERTICAL_SPEED;
        compass = data.MAGNETIC_COMPASS;
        airspeed = data.AIRSPEED_INDICATE;
        latitude = data.LATITUDE;
        longitude = data.LONGITUDE;

        //Fuel
        fuel_percentage = data.FUEL_PERCENTAGE;

        //Autopilot
        autopilot_master = data.AUTOPILOT_MASTER;
        autopilot_nav_selected = data.AUTOPILOT_NAV_SELECTED;
        autopilot_wing_leveler = data.AUTOPILOT_WING_LEVELER;
        autopilot_heading_lock = data.AUTOPILOT_HEADING_LOCK;
        autopilot_heading_lock_dir = data.AUTOPILOT_HEADING_LOCK_DIR;
        autopilot_altitude_lock = data.AUTOPILOT_ALTITUDE_LOCK;
        autopilot_altitude_lock_var = data.AUTOPILOT_ALTITUDE_LOCK_VAR;
        autopilot_attitude_hold = data.AUTOPILOT_ATTITUDE_HOLD;
        autopilot_glidescope_hold = data.AUTOPILOT_GLIDESLOPE_HOLD;
        autopilot_approach_hold = data.AUTOPILOT_APPROACH_HOLD;
        autopilot_backcourse_hold = data.AUTOPILOT_BACKCOURSE_HOLD;
        autopilot_vertical_hold = data.AUTOPILOT_VERTICAL_HOLD
        autopilot_vertical_hold_var = data.AUTOPILOT_VERTICAL_HOLD_VAR;
        autopilot_pitch_hold = data.AUTOPILOT_PITCH_HOLD;
        autopilot_pitch_hold_ref = data.AUTOPILOT_PITCH_HOLD_REF;
        autopilot_flight_director_active = data.AUTOPILOT_FLIGHT_DIRECTOR_ACTIVE;
        autopilot_airspeed_hold = data.AUTOPILOT_AIRSPEED_HOLD;
        autopilot_airspeed_hold_var = data.AUTOPILOT_AIRSPEED_HOLD_VAR;

        //Control surfaces
        gear_handle_position = data.GEAR_HANDLE_POSITION;
        elevator_trim_pct = data.ELEVATOR_TRIM_PCT;
        elevator_trim_pct_reversed = - elevator_trim_pct
        //rudder_trim_pct = data.RUDDER_TRIM_PCT;
        flaps_handle_pct = data.FLAPS_HANDLE_PERCENT;
        flaps_handle_pct_reversed = - flaps_handle_pct;

        //Cabin
        cabin_no_smoking_alert_switch = data.CABIN_NO_SMOKING_ALERT_SWITCH;
        cabin_seatbelts_alert_switch = data.CABIN_SEATBELTS_ALERT_SWITCH;

    });
    return false;
}


function displayData() {
    //Navigation
    $("#altitude").text(altitude);
    $("#compass").text(compass);
    $("#vertical-speed").text(vertical_speed);
    $("#airspeed").text(airspeed);

    //Fuel
    $("#fuel-percentage").text(fuel_percentage);
    $("#fuel-percentage-bar").css("width", fuel_percentage+"%");

    //Autopilot
    checkAndUpdateButton("#autopilot-master", autopilot_master, "Engaged", "Disengaged");
    checkAndUpdateButton("#autopilot-wing-leveler", autopilot_wing_leveler);
    checkAndUpdateButton("#autopilot-heading-lock", autopilot_heading_lock);
    checkAndUpdateButton("#autopilot-altitude-lock", autopilot_altitude_lock);
    checkAndUpdateButton("#autopilot-airspeed-hold", autopilot_airspeed_hold);
    checkAndUpdateButton("#autopilot-attitude-hold", autopilot_attitude_hold);
    checkAndUpdateButton("#autopilot-backcourse-hold", autopilot_backcourse_hold);
    checkAndUpdateButton("#autopilot-approach-hold", autopilot_approach_hold);
    checkAndUpdateButton("#autopilot-vertical-hold", autopilot_vertical_hold);

    $("#autopilot-heading-lock-dir").attr('placeholder', autopilot_heading_lock_dir);
    $("#autopilot-altitude-lock-var").attr('placeholder', autopilot_altitude_lock_var);
    $("#autopilot-airspeed-hold-var").attr('placeholder', autopilot_airspeed_hold_var);
    $("#autopilot-pitch-hold-ref").attr('placeholder', autopilot_pitch_hold_ref);
    $("#autopilot-vertical-hold-ref").attr('placeholder', autopilot_vertical_hold_var);

    //Control surfaces
    $("#gear-handle-position").html(gear_handle_position);
    if (gear_handle_position === "UP"){
        $("#gear-handle-position").removeClass("btn-success").addClass("btn-danger");
    } else {
        $("#gear-handle-position").removeClass("btn-danger").addClass("btn-success");
    }

    $("#flaps-handle-pct").text(flaps_handle_pct);
    $("#flaps-slider").slider({values: [flaps_handle_pct_reversed]})

    $("#elevator-trim-pct").text(elevator_trim_pct);
    $("#elevator-trim-slider").slider({values: [elevator_trim_pct_reversed]})

    //$("#rudder-trim-pct").text(rudder_trim_pct);
    //$("#rudder-trim-slider").slider({values: [rudder_trim_pct]})

    //Cabin
    if (cabin_seatbelts_alert_switch === 1){
        $("#seatbelt-sign").removeClass("btn-outline-danger").addClass("btn-danger").html("Seatbelt sign on");
    } else {
        $("#seatbelt-sign").removeClass("btn-danger").addClass("btn-outline-danger").html("Seatbelt sign off");
    }

    if (cabin_no_smoking_alert_switch === 1){
        $("#no-smoking-sign").removeClass("btn-outline-danger").addClass("btn-danger").html("No smoking sign on");
    } else {
        $("#no-smoking-sign").removeClass("btn-danger").addClass("btn-outline-danger").html("No smoking sign off");
    }}

function checkAndUpdateButton(buttonName, variableToCheck, onText="On", offText="Off") {
    if (variableToCheck === 1) {
        $(buttonName).removeClass("btn-danger").addClass("btn-success").html(onText);
    } else {
        $(buttonName).removeClass("btn-success").addClass("btn-danger").html(offText);
    }
}


function toggleFollowPlane() {
    followPlane = !followPlane;
    if (followPlane === true) {
        $("#followMode").text("Moving map enabled")
        $("#followModeButton").removeClass("btn-outline-danger").addClass("btn-primary")
    }
    if (followPlane === false) {
        $("#followMode").text("Moving map disabled")
        $("#followModeButton").removeClass("btn-primary").addClass("btn-outline-danger")
    }
}

function updateMap() {
    var pos = L.latLng(latitude, longitude);

    marker.slideTo(	pos, {
        duration: 1500,
    });
    marker.setRotationAngle(compass);

    if (followPlane === true) {
        map.panTo(pos);
    }
}

function setSimDatapoint(datapointToSet, valueToUse) {
    url_to_call = "/datapoint/"+datapointToSet+"/set";
    $.post( url_to_call, { value_to_use: valueToUse } );
}

function triggerSimEvent(eventToTrigger, valueToUse, hideAlert = false){
    url_to_call = "/event/"+eventToTrigger+"/trigger";
    $.post( url_to_call, { value_to_use: valueToUse } );

    if (!hideAlert) {
        temporaryAlert('', "Sending instruction", "success")
    }
}

function triggerSimEventFromField(eventToTrigger, fieldToUse, messageToDisplay = null){
    // Get the field and the value in there
    fieldToUse = "#" + fieldToUse
    valueToUse = $(fieldToUse).val();

    // Pass it to the API
    url_to_call = "/event/"+eventToTrigger+"/trigger";
    $.post( url_to_call, { value_to_use: valueToUse } );

    // Clear the field so it can be repopulated with the placeholder
    $(fieldToUse).val("")

    if (messageToDisplay) {
        temporaryAlert('', messageToDisplay + " to " + valueToUse, "success")
    }

}

function triggerCustomEmergency(emergency_type) {
    url_to_call = "/custom_emergency/" + emergency_type
    $.post (url_to_call)

    if (emergency_type === "random_engine_fire") {
        temporaryAlert("Fire!", "Random engine fire trigger sent", "error")
    }
}


function temporaryAlert(title, message, icon) {
    let timerInterval

    Swal.fire({
        title: title,
        html: message,
        icon: icon,
        timer: 2000,
        timerProgressBar: true,
        onBeforeOpen: () => {
            Swal.showLoading()
            timerInterval = setInterval(() => {
                const content = Swal.getContent()
                if (content) {
                    const b = content.querySelector('b')
                    if (b) {
                        b.textContent = Swal.getTimerLeft()
                    }
                }
            }, 100)
        },
        onClose: () => {
            clearInterval(timerInterval)
        }
    }).then((result) => {
        /* Read more about handling dismissals below */
        if (result.dismiss === Swal.DismissReason.timer) {
            console.log('I was closed by the timer')
        }
    })
}

msfs_landing_inspector.py
from flask import Flask, render_template, jsonify, request
from SimConnectCust import *
from threading import Thread
import time

simconnect_dict = {}

# Flask WebApp
def flask_thread_func(threadname):
    
    global simconnect_dict
    
    app = Flask(__name__)
    
    @app.route('/_stuff', methods = ['GET'])
    def stuff():
        global simconnect_dict
        
        return jsonify(simconnect_dict)

    @app.route('/')
    def index():
        return render_template('body.html')

    app.run(host='0.0.0.0', debug=False, use_reloader=False)
    
    

# SimConnect  App
def simconnect_thread_func(threadname):
    
    global simconnect_dict
    
    # Init SimConnect
    sm = SimConnect()
    aq = AircraftRequests(sm, _time = 0, _attemps = 3)
    
    # Init variables
    run_app = 1
    airborne = False
    g_force = 0
    v_speed = 0
    plane_alt_above_ground = 0
    sim_on_ground = 0
    g_force_prev = 0
    v_speed_prev = 0
    plane_alt_above_ground_prev = 0
    sim_on_ground_prev = 0
    v_speed_list_all = []
    g_force_list_all = []
    v_speed_list = []
    g_force_list = []
    plane_alt_above_ground_list = []
    v_speed_list_ground = []
    g_force_list_ground = []
    plane_alt_above_ground_list_ground = []
    sim_on_ground_list = [1,1,1,1,1,1,1,1]
    sim_on_ground_list_all = []
    airspeed = 0
    airspeed_prev = 0
    airspeed_list = []
    airspeed_list_ground = []
    pitch = 0
    pitch_prev = 0
    pitch_list = []
    pitch_list_ground = []
    bank = 0
    bank_prev = 0
    bank_list = []
    bank_list_ground = []
    heading = 0
    heading_prev = 0
    heading_list = []
    heading_list_ground = []
    wind_speed = 0
    wind_speed_prev = 0
    wind_speed_list = []
    wind_speed_list_ground = []
    wind_direction = 0
    wind_direction_prev = 0
    wind_direction_list = []
    wind_direction_list_ground = []
    airspeed_list_all = []
    pitch_list_all = []
    bank_list_all = []
    heading_list_all = []
    wind_speed_list_all = []
    wind_direction_list_all = []
    
    simconnect_dict["G_FORCE"] = 0
    simconnect_dict["VERTICAL_SPEED"] = 0
    simconnect_dict["SIM_ON_GROUND"] = 0
    simconnect_dict["G_FORCE_LANDING"] = "N/A"
    simconnect_dict["VERTICAL_SPEED_LANDING"] = "N/A"
    simconnect_dict["SIM_ON_GROUND_LIST"] = "N/A"
    simconnect_dict["AIRBORNE"] = 0
    simconnect_dict["G_FORCE_LIST"] = g_force_list
    simconnect_dict["V_SPEED_LIST"] = v_speed_list
    simconnect_dict["G_FORCE_LANDING_LIST"] = "N/A"
    simconnect_dict["G_FORCE_LANDING_LIST_GROUND"] = "N/A"
    simconnect_dict["VERTICAL_SPEED_LANDING_LIST"] = "N/A"    
    simconnect_dict["VERTICAL_SPEED_LANDING_LIST_GROUND"] = "N/A"
    simconnect_dict["PLANE_ALT_ABOVE_GROUND_LIST"] = plane_alt_above_ground_list
    simconnect_dict["PLANE_ALT_ABOVE_GROUND_LIST_GROUND"] = plane_alt_above_ground_list_ground
    simconnect_dict["LANDING_RATING"] = "N/A"
    simconnect_dict["AIRSPEED"] = "N/A"
    simconnect_dict["AIRSPEED_LIST"] = airspeed_list
    simconnect_dict["AIRSPEED_LIST_GROUND"] = airspeed_list_ground
    simconnect_dict["PITCH"] = "N/A"
    simconnect_dict["PITCH_LIST"] = pitch_list
    simconnect_dict["PITCH_LIST_GROUND"] = pitch_list_ground
    simconnect_dict["BANK"] = "N/A"
    simconnect_dict["BANK_LIST"] = bank_list
    simconnect_dict["BANK_LIST_GROUND"] = bank_list_ground
    simconnect_dict["HEADING"] = "N/A"
    simconnect_dict["HEADING_LIST"] = heading_list
    simconnect_dict["HEADING_LIST_GROUND"] = heading_list_ground
    simconnect_dict["WIND_SPEED"] = "N/A"
    simconnect_dict["WIND_SPEED_LIST"] = wind_speed_list
    simconnect_dict["WIND_SPEED_LIST_GROUND"] = wind_speed_list_ground
    simconnect_dict["WIND_DIRECTION"] = "N/A"
    simconnect_dict["WIND_DIRECTION_LIST"] = wind_direction_list
    simconnect_dict["WIND_DIRECTION_LIST_GROUND"] = wind_direction_list_ground
    simconnect_dict["LANDING_COUNTER"] = 0
    
    # Create empty labels for charts
    labels_list = []
    for i in range(150):
        labels_list.append("")
    simconnect_dict["LABELS"] = labels_list
    
    # Run Simconnect Calculations
    while run_app == 1:
            
        # Get Current Data 
        # Fix for -999999 values
        v_speed = round(aq.get("VELOCITY_WORLD_Y")*60)
        if v_speed < -99999:
            v_speed = v_speed_prev
        else:
            v_speed_prev = v_speed

        g_force = round(aq.get("G_FORCE"), 2)
        if g_force < -99999:
            g_force = g_force_prev
        else:
            g_force_prev = g_force
            
        plane_alt_above_ground = round(aq.get("PLANE_ALT_ABOVE_GROUND"), 1)
        if plane_alt_above_ground < -99999:
            plane_alt_above_ground = plane_alt_above_ground_prev
        else:
            plane_alt_above_ground_prev = plane_alt_above_ground
        
        sim_on_ground = aq.get("SIM_ON_GROUND")
        if sim_on_ground < -99999:
            sim_on_ground = sim_on_ground_prev
        else:
            sim_on_ground_prev = sim_on_ground        
               
        airspeed = round(aq.get("AIRSPEED_TRUE"))
        if airspeed < -99999:
            airspeed = airspeed_prev
        else:
            airspeed_prev = airspeed
            
        pitch = round(aq.get("PLANE_PITCH_DEGREES") * 180 / 3.1416, 1) * (-1)
        if abs(pitch) > 999:
            pitch = pitch_prev
        else:
            pitch_prev = pitch 

        bank = round((aq.get("PLANE_BANK_DEGREES") * 180 / 3.1416), 1)
        if abs(bank) > 999:
            bank = bank_prev
        else:
            bank_prev = bank

        heading = round(aq.get("PLANE_HEADING_DEGREES_TRUE") * 180 / 3.1416)
        if abs(heading) > 999:
            heading = heading_prev
        else:
            heading_prev = heading
            
        wind_speed = round(aq.get("AMBIENT_WIND_VELOCITY"), 1)
        if wind_speed < -99999:
            wind_speed = wind_speed_prev
        else:
            wind_speed_prev = wind_speed

        wind_direction = round(aq.get("AMBIENT_WIND_DIRECTION"))
        if wind_direction < -99999:
            wind_direction = wind_direction_prev
        else:
            wind_direction_prev = wind_direction
        
        # Make lists
        sim_on_ground_list.insert(0, sim_on_ground)
        if len(sim_on_ground_list) > 21:
           sim_on_ground_list.pop()
           
        sim_on_ground_list_all.insert(0, sim_on_ground)
        if len(sim_on_ground_list_all) > 151:
           sim_on_ground_list_all.pop()

        v_speed_list_all.insert(0, v_speed)
        if len(v_speed_list_all) > 151:
            v_speed_list_all.pop()
        g_force_list_all.insert(0, g_force)
        if len(g_force_list_all) > 151:
            g_force_list_all.pop()
        airspeed_list_all.insert(0, airspeed)
        if len(airspeed_list_all) > 151:
            airspeed_list_all.pop() 
        pitch_list_all.insert(0, pitch)
        if len(pitch_list_all) > 151:
            pitch_list_all.pop() 
        bank_list_all.insert(0, bank)
        if len(bank_list_all) > 151:
            bank_list_all.pop() 
        heading_list_all.insert(0, heading)
        if len(heading_list_all) > 151:
            heading_list_all.pop() 
        wind_speed_list_all.insert(0, wind_speed)
        if len(wind_speed_list_all) > 151:
            wind_speed_list_all.pop()
        wind_direction_list_all.insert(0, wind_direction)
        if len(wind_direction_list_all) > 151:
            wind_direction_list_all.pop()
        
        
        # Make lists for graph - separation between airborne and landing
        if sim_on_ground == 1:
            v_speed_list.insert(0, "null")
            if len(v_speed_list) > 151:
                v_speed_list.pop() 
            g_force_list.insert(0, "null")
            if len(g_force_list) > 151:
               g_force_list.pop()
            plane_alt_above_ground_list.insert(0, "null")
            if len(plane_alt_above_ground_list) > 151:
               plane_alt_above_ground_list.pop()
            v_speed_list_ground.insert(0, v_speed)
            if len(v_speed_list_ground) > 151:
                v_speed_list_ground.pop() 
            g_force_list_ground.insert(0, g_force)
            if len(g_force_list_ground) > 151:
               g_force_list_ground.pop()
            plane_alt_above_ground_list_ground.insert(0, plane_alt_above_ground)
            if len(plane_alt_above_ground_list_ground) > 151:
               plane_alt_above_ground_list_ground.pop()
            airspeed_list.insert(0, "null")
            if len(airspeed_list) > 151:
               airspeed_list.pop()
            airspeed_list_ground.insert(0, airspeed)
            if len(airspeed_list_ground) > 151:
               airspeed_list_ground.pop()
            pitch_list.insert(0, "null")
            if len(pitch_list) > 151:
               pitch_list.pop()
            pitch_list_ground.insert(0, pitch)
            if len(pitch_list_ground) > 151:
               pitch_list_ground.pop()
            bank_list.insert(0, "null")
            if len(bank_list) > 151:
               bank_list.pop()
            bank_list_ground.insert(0, bank)
            if len(bank_list_ground) > 151:
               bank_list_ground.pop()
            heading_list.insert(0, "null")
            if len(heading_list) > 151:
               heading_list.pop()
            heading_list_ground.insert(0, heading)
            if len(heading_list_ground) > 151:
               heading_list_ground.pop()
            wind_speed_list.insert(0, "null")
            if len(wind_speed_list) > 151:
               wind_speed_list.pop()
            wind_speed_list_ground.insert(0, wind_speed)
            if len(wind_speed_list_ground) > 151:
               wind_speed_list_ground.pop()
            wind_direction_list.insert(0, "null")
            if len(wind_direction_list) > 151:
               wind_direction_list.pop()
            wind_direction_list_ground.insert(0, wind_direction)
            if len(wind_direction_list_ground) > 151:
               wind_direction_list_ground.pop()
        else:
            v_speed_list.insert(0, v_speed)
            if len(v_speed_list) > 151:
                v_speed_list.pop() 
            g_force_list.insert(0, g_force)
            if len(g_force_list) > 151:
               g_force_list.pop()
            plane_alt_above_ground_list.insert(0, plane_alt_above_ground)
            if len(plane_alt_above_ground_list) > 151:
               plane_alt_above_ground_list.pop()
            v_speed_list_ground.insert(0, "null")
            if len(v_speed_list_ground) > 151:
                v_speed_list_ground.pop() 
            g_force_list_ground.insert(0, "null")
            if len(g_force_list_ground) > 151:
               g_force_list_ground.pop()
            plane_alt_above_ground_list_ground.insert(0, "null")
            if len(plane_alt_above_ground_list_ground) > 151:
               plane_alt_above_ground_list_ground.pop()
            airspeed_list.insert(0, airspeed)
            if len(airspeed_list) > 151:
               airspeed_list.pop()
            airspeed_list_ground.insert(0, "null")
            if len(airspeed_list_ground) > 151:
               airspeed_list_ground.pop()
            pitch_list.insert(0, pitch)
            if len(pitch_list) > 151:
               pitch_list.pop()
            pitch_list_ground.insert(0, "null")
            if len(pitch_list_ground) > 151:
               pitch_list_ground.pop()
            bank_list.insert(0, bank)
            if len(bank_list) > 151:
               bank_list.pop()
            bank_list_ground.insert(0, "null")
            if len(bank_list_ground) > 151:
               bank_list_ground.pop()
            heading_list.insert(0, heading)
            if len(heading_list) > 151:
               heading_list.pop()
            heading_list_ground.insert(0, "null")
            if len(heading_list_ground) > 151:
               heading_list_ground.pop()
            wind_speed_list.insert(0, wind_speed)
            if len(wind_speed_list) > 151:
               wind_speed_list.pop()
            wind_speed_list_ground.insert(0, "null")
            if len(wind_speed_list_ground) > 151:
               wind_speed_list_ground.pop()
            wind_direction_list.insert(0, wind_direction)
            if len(wind_direction_list) > 151:
               wind_direction_list.pop()
            wind_direction_list_ground.insert(0, "null")
            if len(wind_direction_list_ground) > 151:
               wind_direction_list_ground.pop()

        
        # Populate vars to JSON dictionary
        simconnect_dict["G_FORCE"] = g_force
        simconnect_dict["VERTICAL_SPEED"] = v_speed
        simconnect_dict["SIM_ON_GROUND"] = sim_on_ground
        simconnect_dict["AIRBORNE"] = airborne

        
        # Make landing/airborne decision
        if airborne == True and sum(sim_on_ground_list) == 20:
            # Fix - need to get the last value before on ground readings
            v_speed_list_touchdown = v_speed_list_ground
            g_force_list_touchdown = g_force_list_ground
            change_last = False
            for idx, element in enumerate(v_speed_list_touchdown):
                if idx >= 1:
                    if element == "null" and sim_on_ground_list_all[idx-1] == 1:
                        if change_last == False:
                            v_speed_list_touchdown[idx] = v_speed_list[idx]
                            g_force_list_touchdown[idx] = g_force_list[idx]
                            change_last = True
                        else:
                            change_last = False
                    else:
                        change_last = False
            
            v_speed_list_touchdown = [0 if x=="null" else x for x in v_speed_list_touchdown]
            g_force_list_touchdown = [0 if x=="null" else x for x in g_force_list_touchdown]
            
            # Get Values for Touchdown 
            simconnect_dict["VERTICAL_SPEED_LANDING"] = min(v_speed_list_touchdown)
            
            # Calculate Custom G-Force based on vertical speed
            # Model uses Harvsine acceleration model for peak acceleration https://www.nhtsa.gov/sites/nhtsa.dot.gov/files/18esv-000501.pdf
            # For time a custom function is used that ranges from 0.25 for GA to 0.35 for airliners. This simulates the rigidness of suspention.
            custom_g_force_v_speed_m_per_s = (min(v_speed_list_touchdown) * (-0.3048) / 60)
            plane_weight = max(round(aq.get("TOTAL_WEIGHT") * 0.45), 50)
            custom_g_force_impact_duration = 0.355 + (-0.103/(1 + (plane_weight/15463)**1.28))
            custom_g_force = 1 + (((2 * custom_g_force_v_speed_m_per_s) / custom_g_force_impact_duration) / 9.80665)
            custom_g_force = round(custom_g_force, 2)

            # Get Index of Touchdown from min Vertical Speed
            index_vspeed_touchdown = v_speed_list_touchdown.index(min(v_speed_list_touchdown))
            
            # Check if SimConnect G force is within +/- 10% of custom G force. If not use custom G force
            if g_force_list_touchdown[index_vspeed_touchdown] < custom_g_force*0.9 or g_force_list_touchdown[index_vspeed_touchdown] > custom_g_force*1.1:
                g_force_list_touchdown[index_vspeed_touchdown] = custom_g_force
                g_force_list[index_vspeed_touchdown] = custom_g_force
                g_force_list_ground[index_vspeed_touchdown] = custom_g_force
            
            # Get the rest of touchdown data based on vertical speed index
            simconnect_dict["G_FORCE_LANDING"] = max(g_force_list_touchdown)
            simconnect_dict["AIRSPEED"] = airspeed_list_all[index_vspeed_touchdown]
            simconnect_dict["PITCH"] = pitch_list_all[index_vspeed_touchdown]
            simconnect_dict["BANK"] = bank_list_all[index_vspeed_touchdown]
            simconnect_dict["HEADING"] = heading_list_all[index_vspeed_touchdown]
            simconnect_dict["WIND_SPEED"] = wind_speed_list_all[index_vspeed_touchdown]
            simconnect_dict["WIND_DIRECTION"] = wind_direction_list_all[index_vspeed_touchdown]
            
            # Create Lists for Graphs
            simconnect_dict["G_FORCE_LANDING_LIST"] = g_force_list[::-1]*1
            simconnect_dict["G_FORCE_LANDING_LIST_GROUND"] = g_force_list_ground[::-1]*1
            v_speed_list_neg = [elem * (-1) if elem != "null" else "null" for elem in v_speed_list]
            v_speed_list_ground_neg = [elem * (-1) if elem != "null" else "null" for elem in v_speed_list_ground]
            simconnect_dict["VERTICAL_SPEED_LANDING_LIST"] = v_speed_list_neg[::-1]*1
            simconnect_dict["VERTICAL_SPEED_LANDING_LIST_GROUND"] = v_speed_list_ground_neg[::-1]*1
            simconnect_dict["PLANE_ALT_ABOVE_GROUND_LIST"] = plane_alt_above_ground_list[::-1]*1
            simconnect_dict["PLANE_ALT_ABOVE_GROUND_LIST_GROUND"] = plane_alt_above_ground_list_ground[::-1]*1
            simconnect_dict["LANDING_COUNTER"] = simconnect_dict["LANDING_COUNTER"] + 1            
            simconnect_dict["AIRSPEED_LIST"] = airspeed_list[::-1]*1
            simconnect_dict["AIRSPEED_LIST_GROUND"] = airspeed_list_ground[::-1]*1
            simconnect_dict["PITCH_LIST"] = pitch_list[::-1]*1
            simconnect_dict["PITCH_LIST_GROUND"] = pitch_list_ground[::-1]*1
            simconnect_dict["BANK_LIST"] = bank_list[::-1]*1
            simconnect_dict["BANK_LIST_GROUND"] = bank_list_ground[::-1]*1
            simconnect_dict["HEADING_LIST"] = heading_list[::-1]*1
            simconnect_dict["HEADING_LIST_GROUND"] = heading_list_ground[::-1]*1
            simconnect_dict["WIND_SPEED_LIST"] = wind_speed_list[::-1]*1
            simconnect_dict["WIND_SPEED_LIST_GROUND"] = wind_speed_list_ground[::-1]*1
            simconnect_dict["WIND_DIRECTION_LIST"] = wind_direction_list[::-1]*1
            simconnect_dict["WIND_DIRECTION_LIST_GROUND"] = wind_direction_list_ground[::-1]*1
            simconnect_dict["SIM_ON_GROUND_LIST"] = sim_on_ground_list_all[::-1]*1
            
            # Landing Rating Based on G-Forces
            if simconnect_dict["VERTICAL_SPEED_LANDING"] > -60:
                simconnect_dict["LANDING_RATING"] = "Very soft landing"
            elif simconnect_dict["VERTICAL_SPEED_LANDING"] > -120:
                simconnect_dict["LANDING_RATING"] = "Soft landing"
            elif simconnect_dict["VERTICAL_SPEED_LANDING"] > -200:
                simconnect_dict["LANDING_RATING"] = "Average landing"
            elif simconnect_dict["VERTICAL_SPEED_LANDING"] > -300:
                simconnect_dict["LANDING_RATING"] = "Firm landing"
            elif simconnect_dict["VERTICAL_SPEED_LANDING"] > -400:
                simconnect_dict["LANDING_RATING"] = "Hard landing"
            elif simconnect_dict["VERTICAL_SPEED_LANDING"] > -600:
                simconnect_dict["LANDING_RATING"] = "Very hard landing"
            else:
                simconnect_dict["LANDING_RATING"] = "Structural damage to plane"
            airborne = False
            
            
        # Decision for Airborne
        if sum(sim_on_ground_list) == 0 and airborne == False:
            airborne = True

        
if __name__ == "__main__":
    thread1 = Thread(target = simconnect_thread_func, args=('Thread-1', ))
    thread2 = Thread(target = flask_thread_func, args=('Thread-2', ))
    thread1.start()
    thread2.start()
    
    time.sleep(.5)
    
    print("\n*********************")
    print("MSFS Landing Inspector initialized")
    print("Load 'localhost:5000' in your browser to start MSFS Landing Inspector")
    print("Tip: You can access MSFS Landing Inspector from your mobile device. Check install_instructions.txt for instructions.")
        